File: nexarq_cli/agents/builtin/logging_agent.py
"""Logging agent – log level usage, PII in logs, structured logging."""
from nexarq_cli.agents.base import BaseAgent, AgentPermissions, Severity

_PROMPT = """\
You are a logging and observability expert.

Analyze the following {language} code diff for logging quality:
- Personal Identifiable Information (PII) logged (names, emails, passwords, tokens)
- Secrets or credentials in log messages
- Missing logging for critical operations (auth events, data mutations)
- Wrong log level usage (debug in production, error for expected conditions)
- Unstructured logging where structured (JSON) logging should be used
- Log messages without context (no request ID, user ID, correlation ID)
- print() statements left in production code instead of proper logging
- Missing log rotation or size limits configuration
- Excessive logging that could impact performance
- Missing audit trail for compliance-sensitive operations

For each issue:
  SEVERITY: [CRITICAL|HIGH|MEDIUM|LOW]
  LOCATION: <code snippet>
  ISSUE: <description>
  FIX: <corrected logging>

If logging is correct: "Logging practices look appropriate."

Code diff:
```{language}
{diff}
```"""


class LoggingAgent(BaseAgent):
    name = "logging"
    description = "PII in logs, log levels, structured logging, audit trail"
    severity = Severity.HIGH

    def __init__(self) -> None:
        super().__init__()
        self.permissions = AgentPermissions(read_diff_only=True)

    def build_prompt(self, diff: str, language: str, context: dict | None = None) -> str:
        return _PROMPT.format(language=language, diff=diff)

File: agents/docs/logging-agent.ts
import type { AgentDefinition } from '@nexarq/common/interfaces'
import { SHARED_SYSTEM_PREFIX, buildUserPrompt } from '../agent-template.ts'

const instructions = `Focus ONLY on logging quality and privacy issues in this diff.

Check for:
- PII or sensitive data logged (passwords, tokens, email addresses, phone numbers)
- Logging at incorrect levels (using console.log in production code, DEBUG in hot paths)
- Missing error logging in catch blocks
- Logs that include raw request/response bodies containing user data
- Excessive logging in tight loops (performance impact)
- Log messages that are too vague to be actionable for debugging
- Missing structured logging fields (no correlation ID, no request context)`

export const loggingAgent: AgentDefinition = {
  name: 'logging',
  displayName: 'Logging',
  description: 'PII in logs, incorrect log levels, missing error logs',
  severity: 'medium',
  tier: 2,
  needsTools: false,
  systemPrompt: SHARED_SYSTEM_PREFIX,
  buildPrompt: (diff, language, context) => buildUserPrompt(instructions, diff, language, context),
}
